### ISSUE TITLE
Session locking

### ISSUE THREAD

**Issue by michael-grunder** [phpredis/phpredis#1267]
Creating this issue to track session locking functionality added by @SkydiveMarius.

The branch with locking functionality can be found [here](https://github.com/phpredis/phpredis/tree/session-locking)

Session locking is now possible for the Redis session handler by enabling it in your PHP INI file.  Original description of how that works [here](https://github.com/phpredis/phpredis/pull/1181#issue-228763229).

It's probably a good idea to implement this functionality in the Redis Cluster session handler as well so phpredis is consistent.  I will try to get that implemented in the next days.

If people who want to use the locking functionality would test the new features that would be great.  Neither myself or @yatsukhnenko use phpredis to store session information so our ability to test in a real world like way is limited.

Cheers,
Mike

**Comment by mrmark** [comment]
Thanks for doing this, this will be really wonderful.  I saw you are releasing the lock in `PS_CLOSE_FUNC`, does that auto-release the lock when the process closes?  That's something that I have had problems with in user land session locking, some cases you cannot release the lock because PHP just stopped running user code.  Didn't see this problem though with extensions that handled their own session locking, like memcached.

+1 for clusters, we are using that functionality.

**Comment by SkydiveMarius** [comment]
Thanks for the positive feedback, credits also for @KiNgMaR for his support during development.
@mrmark Would be nice if you could support us by testing the feature in similar environments as production.

As @michael-grunder already mentioned, a TTL is defined for the locking key (*_LOCK) in Redis to release the lock if the process dies unexpectedly.
By default TTL is equal to may_execution_time, but can be overwritten by the following php.ini setting:
```ini
; Setting the lock TTL to 1 minute
redis.session.lock_expire=60
```

**Comment by mrmark** [comment]
> I believe that unless PHP actually crashes (segmentation fault or similar) it will always call PS_CLOSE_FUNC during shutdown

This sounds good.  My particular use case was with Nginx killing the process if it took too long.  I'm hoping `PS_CLOSE_FUNC` is the cure.

> Would be nice if you could support us by testing the feature in similar environments as production.

I'll throw it onto my todo list, but I'll likely wait until cluster support is ready since we use both.

Cheers and thanks all!

**Comment by asharpe** [comment]
Another thanks for doing this, much appreciated :)

I'd just like to confirm the expected behaviour when not being able to acquire the lock, and I'll give it a go by walking through the code as I see it.

If we fail to acquire the lock at session startup (https://github.com/phpredis/phpredis/blob/session-locking/redis_session.c#L613) there will still be a request to get the session data.  Assuming this succeeds, the process is running with potentially invalid session data.

Question: is there a way for userland to know it's in this situation?

Continuing on as the process that failed to acquire the lock, if it comes time to write something to the session we check for write access (https://github.com/phpredis/phpredis/blob/session-locking/redis_session.c#L682) and fail if we don't have it.  This is another place that userland might like to know something went wrong, but I don't think the PHP API allows for that (http://php.net/manual/en/function.session-write-close.php).

Is there a description of the semantics around locking that I haven't found yet?

Thanks again for implementing this feature :)

**Comment by asharpe** [comment]
I'm using php 5.4 on scientific linux 6 and am having trouble with this branch and `session_regenerate_id(true)`.  Should this be expected to work?

The tests all pass during the build.

**Comment by SkydiveMarius** [comment]
Thanks for reporting this issue.
Im in holiday until 07.01. I will have a look as soon I'm back.

**Comment by neerolyte** [comment]
The current defaults for [lock_wait_time (2000 usecs)](https://github.com/phpredis/phpredis/blob/5a3f762/redis_session.c#L276) and [lock_retries (10)](https://github.com/phpredis/phpredis/blob/5a3f762/redis_session.c#L282) seem excessively small - out of the box phpredis is only waiting a total of 20ms before giving up on the lock and using the session as is. The [default timeout for the memcache extension is 15 seconds](http://php.net/manual/en/memcache.ini.php) - 750x larger.

Would you consider raising the defaults to something like a lock_wait_time of 20000 usecs and a lock_retries of 750 to match? I expect this would simplify transition between modules a bit.

**Comment by michael-grunder** [comment]
I think you're probably right that 2000 usecs is to small for the default retry wait time and should _probably_ be some number of milliseconds (maybe 50 or 100 by default?).  I'm guessing a 2 millisecond delay is effectively zero (will get lost in the normal network latency).

**Comment by SkydiveMarius** [comment]
I also agree with @neerolyte, as the current default values, have been not well calculated.
In our test environment we increased the value to lock_wait_time to 50ms with good results.
Maybe we should think about disabling lock_retries per default, as most users are already covered with TTL=max_execution_time.

**Comment by SkydiveMarius** [comment]
I think we found another bug in combination with session_regenerate_id() and [SessionHandler class](https://secure.php.net/manual/it/class.sessionhandler.php).
The bug should affect all users of the [HttpFoundation NativeSessionStorage](http-foundation/Session/Storage/NativeSessionStorage.php).

## Summary
1. Register the native SessionHandler as session wrapper
2. session_regenerate_id()
3. It seems that PHP internally neither calls PS_CREATE_SID nor PS_OPEN. Instead directly PS_WRITE is called. PS_WRITE tries to use the old session id stored in the pool (lock_status->session_key) and fails

## Details/Reproducing
````php
class TestHandler implements SessionHandlerInterface
{
    /**
     * @var SessionHandler
     */
    private $handler;

    public function __construct()
    {
        $this->handler = new SessionHandler();
    }

    public function close()
    {
        echo "--> TestHandler->close() called\n";
        return $this->handler->close();
    }

    public function destroy($session_id)
    {
        echo "--> TestHandler->destroy() called\n";
        return $this->handler->destroy($session_id);
    }

    public function gc($maxlifetime)
    {
        echo "--> TestHandler->gc() called\n";
        return $this->handler->gc($maxlifetime);
    }

    public function open($save_path, $name)
    {
        echo "--> TestHandler->open() called\n";
        return $this->handler->open($save_path, $name);
    }

    public function read($session_id)
    {
        echo "--> TestHandler->read() called\n";
        return $this->handler->read($session_id);
    }

    public function write($session_id, $session_data)
    {
        echo "--> TestHandler->write() called\n";
        return $this->handler->write($session_id, $session_data);
    }
}

ob_start();

$handler = new TestHandler();
session_set_save_handler($handler);

echo "Calling session_id\n";
session_id('oldSessionId');
echo "Calling session_start\n";
session_start();
echo "Calling session_regenerate_id\n";
session_regenerate_id(true);
echo "Calling session_write_close\n";
session_write_close();

echo "ID: " . session_id() . "\n";
ob_flush();
````

Resulting in the following execution path:
````
Calling session_id
Calling session_start
--> TestHandler->open() called
--> TestHandler->read() called
Calling session_regenerate_id
--> TestHandler->destroy() called
Calling session_write_close
--> TestHandler->write() called
--> TestHandler->close() called
ID: kdaii0i6eo6r1h9r95osuhdo5s36a63ue0a1saba71re1
````

The question is, why PHP internally does not call PS_CREATE_SID when using [SessionHandler class](https://secure.php.net/manual/it/class.sessionhandler.php)?

If this is the correct behavior then I see now only two possible ways:
1.  Check on each read + write if the session key changed
2. Remove session_id from pool, which requires changing the socket/pool determination logic(redis_pool_get_sock()), which is currently based on the session key

@michael-grunder @yatsukhnenko What do you think?

**Comment by yatsukhnenko** [comment]
IMO when somebody wants to implement SessionHandlerInterface he will use Redis object to directly get/set session data from server

**Comment by mvrhov** [comment]
Yeah, they will if they don't care about session locking. And not caring can be a big problem when your application grows.

**Comment by SkydiveMarius** [comment]
I think we should try to fix the issue, as we would **break all Symfony users with PHP5**.
Anyway this might be a good reason for rethinking the current construction, with storing the session_key in the pool.

The only reason for doing this, is that we need the key during session_close to release the lock. Unfortunately the key is not available by macro.
One more suggestion:
1. Remove session_key from pool->lock_status
2. Try to use key provided as macro parameter if possible
3. If Key is not available, fetch it from the PHP engine, if necessary by calling session_id()

What do you think? That would relieve use by avoiding many key saving logic including estrdup() and efree().

**Comment by SkydiveMarius** [comment]
Ok, after trying to implement my last suggestion, the solution seems not that good anymore.

**Drawback comparison**
Removement of session_key from pool->lock_status:
* Lock key is not cached anymore and needs to be recalculcated on each operation
* Another network round trip for fetching the session key in PS_CLOSE() required
 
Check on each write if the session key changed
* Overhead of string compare on each write, and this is only required for PHP < 7

So after many internal discussions (thanks to @KiNgMaR) we came to the point that my first suggested solution (Check on each write if key changed) seems to be the easiest and the best.

I will prepare and create a PR tomorrow.

**Comment by yatsukhnenko** [comment]
@michael-grunder I think it's time to merge session related changes into `develop` so more users will test them and will give us feedback

**Comment by Incognito** [comment]
Because this is a complex issue and a lot of discussion and closed issues/PRs were created over the years, the summary is: Single-instance redis server session locking is now supported and documented in the main README file.

**Comment by Lathanderjk** [comment]
Any news about session locking in cluster?
